File: Problems/hacker_rank/Arrays_and_Sorting/20141226_Insertion_Sort_Part1/solution.c
#include <stdio.h>
#include <stdlib.h>

void print_arr(int* arr, int s) {
  int i = 0;
  printf("%d", *arr);
  for ( i = 1; i < s; ++i ) {
    printf(" %d", *(arr + i));
  }
  printf("\n");
}

void insertion_sort(int* arr, int s)  {
  int V = *(arr + s - 1);
  int idx = s - 2;
  while ( -1 < idx && V < *(arr + idx) )  {
    *(arr + idx + 1) = *(arr + idx);
    print_arr(arr, s);
    --idx;
  }
  *(arr + idx + 1) = V;
  print_arr(arr, s);
}

int main()  {
  int s = 0;
  scanf("%d", &s);
  int* arr = (int*) malloc(sizeof(int) * s);
  int i = 0;
  for ( i = 0; i < s; ++i ) {
    scanf("%d", arr + i);
  }
  insertion_sort(arr, s);
  free(arr);
  return 0;
}

File: Problems/hacker_rank/Arrays_and_Sorting/20141226_Insertion_Sort_Part1/Solution.java
import java.util.Scanner;

public class Solution {
  public static void printArr(final int[] arr)  {
    System.out.print(arr[0]);
    for ( int i = 1; i < arr.length; ++i )  {
      System.out.print(" " + arr[i]);
    }
    System.out.println();
  }

  public static void insertionSort(final int[] arr) {
    int V = arr[arr.length - 1];
    int idx = arr.length - 2;
    while ( -1 < idx && V < arr[idx] )  {
      arr[idx + 1] = arr[idx];
      printArr(arr);
      --idx;
    }
    arr[idx + 1] = V;
    printArr(arr);
  }

  public static void main(final String[] args)  {
    Scanner sc = new Scanner(System.in);
    int s = Integer.parseInt(sc.nextLine());
    String[] inps = sc.nextLine().split(" ");
    int[] arr = new int[inps.length];
    for ( int i = 0; i < inps.length; ++i )
      arr[i] = Integer.parseInt(inps[i]);
    insertionSort(arr);
  }
}
